abn/application/application.go

```diff
@@ -0,0 +1,212 @@
+package application
+
+// application.go - type of application
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/iter8-tools/iter8/base/log"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"sigs.k8s.io/yaml"
+)
+
+const (
+	// KEY is the name of the field in a kubernetes secret
+	KEY = "versionData.yaml"
+)
+
+// Application is an application observed in a kubernetes cluster
+type Application struct {
+	// Name is the value of the app.kubernetes.io/name field
+	Name string
+	// Namespace is the namespace where the application was discovered
+	Namespace string
+	// Tracks is map of tracks identified to versions
+	Tracks
+	// Versions is list of discovered versions
+	Versions
+	// ReaderWriter is used to read/write to persistent storage (a Kubernetes secret)
+	Writer *ApplicationReaderWriter
+}
+
+// Versions is type of version name to versions map
+type Versions map[string]*Version
+
+// Tracks is map of track identifiers to version names
+type Tracks map[string]string
+
+// ApplicationReaderWriter is type used to read/write fromt/to persistent storage
+type ApplicationReaderWriter struct {
+	// Type Kubernetes client
+	Client kubernetes.Interface
+}
+
+// Read reads the application from persistent storage (a Kubernetes secret)
+// - the secret name/namespace is the same as the application
+// - if no application is present in the persistent storage, a new object is created
+func (rw *ApplicationReaderWriter) Read(appName string) (*Application, error) {
```

Comment:
Isn't it simpler to structure the read and write as follows?

```go
// read application from secret (is a yaml file)
 	rawData, ok := secret.Data[KEY]
 	if !ok {
 		log.Logger.Debug("key missing in backing secret")
 		return a, errors.New("secret does not contain expected key: " + KEY)
 	}
 	err = yaml.Unmarshal(rawData, &a)
```

```go
// write application to secret (is a yaml file)
rawData, err := yaml.Marshal(a)
secret.Data[KEY] = rawData
 	if secret.StringData != nil {
 		secret.StringData[KEY] = string(rawData)
 	}
```

This allows you to mirror the in-memory app to the app data in secret -- in the simplest possible way.

Reply:
The app data structure is entirely up to you of course (and along with methods/funcs, can be made to yield any information that is needed).

Reply:
The current approach was designed with the principle of minimizing what gets written to the secret. To this end, only the Versions field is written to the secret. The other fields can be determined from the namespace/name of secret and the event history.

Writing all fields is marginally easier -- it is also the case that when there are no events/metrics the read does not initialize these variables. It is necessary to explicitly check and initialize. Doing so on read makes the rest of the code simpler.